# Velocity adding in special relativity

*This is a [Jupyter Notebook](https://en.wikipedia.org/wiki/Project_Jupyter), a type of document that can contain text, images and runnable Python code. You can view it in your web browser, or transfer a copy to Google Colab with one click and edit it from your device. It can also be downloaded to your PC for more advanced editing.*

*Don't worry about any Python code if you're not a programmer. You can skip over it and just read the results, which are usually tables or graphs. The code is there to help you understand the maths, and to make it easier to experiment with different values.*

*The maths needed for special relativity is not difficult, there's nothing more complicated than high school algebra.*

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/lookbusy1344/Relativity/blob/main/Python/Velocity%20adding.ipynb)

## What happens when you add two speeds, each close to light speed?

Two objects are passing each other, each travelling at 99.99% $c$ (light speed) in opposite directions. What velocity do they measure for each other?

**99.99% $c$ + 99.99% $c$ = ?**

More than light speed? No, heres why.

## In everyday life

In everyday life, we can just add velocities like this:

\begin{equation}
v_{\text{total}} = v_1 + v_2
\end{equation}

So if an object travelling at 100 metres per second passes another object going at 100 m/s in the opposite direction, they measure each other's speed as just:

\begin{equation}
100 + 100 = 200
\end{equation}

But this common sense approach is only an approximation, and breaks down as speeds get closer to $c$.

## Relativistic velocity addition

The correct way to add any velocities is like this:

\begin{equation}
v_{\text{total}} = \frac{v_1 + v_2}{1 + \frac{v_1 v_2}{c^2}}
\end{equation}

The difference is microscopic for everyday speeds, so no one notices. But as you get closer to $c$ it becomes more significant.

The new equation is a bit more complicated, so let's write some Python code to help us. This is where the fun begins!

In [1]:
import math

c = 299_792_458.0  # speed of light, metres per second
c_squared = c**2  # c squared


def add_vel(v1: float, v2: float) -> float:
    if v1 >= c or v2 >= c:
        return math.nan
    return (v1 + v2) / (1.0 + (v1 * v2) / c_squared)


speed1 = 100.0  # speed of first object in metres per second, EDIT THIS
speed2 = 100.0  # speed of second object in metres per second, EDIT THIS
print(f"{speed1} m/s + {speed2} m/s = {add_vel(speed1, speed2)} m/s")

100.0 m/s + 100.0 m/s = 199.99999999997775 m/s


Almost the same answer, to within 10 decimal places. Let's try it with some faster speeds.

In [2]:
from prettytable import PrettyTable, TableStyle
from IPython.display import Markdown

table = PrettyTable()
table.set_style(TableStyle.MARKDOWN)
table.align = "r"
table.field_names = [
    "Velocity1 (m/s)",
    "Velocity2 (m/s)",
    "Added relativistically (m/s)",
    "Discrepancy",
]


def calculate(v: float, round: bool = True) -> None:
    simple = v + v
    relativistic = add_vel(v, v)
    diff = 1.0 - relativistic / simple  # difference as a percentage
    if round:
        rel_str = f"{relativistic:,.1f}"
    else:
        rel_str = f"{relativistic:,}"
    table.add_row([f"{v:,}", f"{v:,}", rel_str, f"{diff:%}"])


# change these, but make sure they are less than light speed
calculate(1_000, False)
calculate(100_000, False)
calculate(1_000_000)
calculate(10_000_000)
calculate(100_000_000)
calculate(200_000_000)

print("Adding velocities relativistically, and discrepancy with newtonian physics")
markdown_table = table.get_string()
display(Markdown(markdown_table));

Adding velocities relativistically, and discrepancy with newtonian physics


| Velocity1 (m/s) | Velocity2 (m/s) | Added relativistically (m/s) | Discrepancy |
|---------------: |---------------: |----------------------------: |-----------: |
|           1,000 |           1,000 |          1,999.9999999777472 |   0.000000% |
|         100,000 |         100,000 |          199,999.97774700134 |   0.000011% |
|       1,000,000 |       1,000,000 |                  1,999,977.7 |   0.001113% |
|      10,000,000 |      10,000,000 |                 19,977,771.7 |   0.111141% |
|     100,000,000 |     100,000,000 |                179,975,072.5 |  10.012464% |
|     200,000,000 |     200,000,000 |                276,805,111.1 |  30.798722% |

## Effect gets proportionally larger

The difference between simple and relativistic addition gets larger as speeds increase. At 2 km/s it is almost zero, but at 200,000 km/s it has shot up to over 30%.

For faster speeds, let's switch to using fractions of $c$ (light speed) instead of m/s. Everything else is the same:

In [3]:
table = PrettyTable()
table.set_style(TableStyle.MARKDOWN)
table.align = "r"
table.field_names = [
    "Velocity1 (c)",
    "Velocity2 (c)",
    "Added relativistically (c)",
    "Discrepancy",
]


def add_vel_c(v1: float, v2: float) -> float:
    # two velocities as fraction of c
    if v1 >= 1.0 or v2 >= 1.0:
        return math.nan
    return (v1 + v2) / (1.0 + (v1 * v2))


def calculate_c(v: float, extra_decimals: bool = False) -> None:
    simple = v + v
    relativistic = add_vel_c(v, v)
    diff = 1.0 - relativistic / simple  # difference as a percentage
    if extra_decimals:
        table.add_row([f"{v:.2%}", f"{v:.2%}", f"{relativistic:.7%}", f"{diff:.1%}"])
    else:
        table.add_row([f"{v:.2%}", f"{v:.2%}", f"{relativistic:.2%}", f"{diff:.1%}"])


# change these, but make sure they are less than light speed, 1.0
calculate_c(0.1)  # 10% light speed
calculate_c(0.2)
calculate_c(0.3)
calculate_c(0.4)
calculate_c(0.5)  # 50% light speed
calculate_c(0.7)
calculate_c(0.9)
calculate_c(0.95)
calculate_c(0.99, extra_decimals=True)
calculate_c(0.9999, extra_decimals=True)  # the original question, 99.99% light speed

print("Adding velocities relativistically, and discrepancy with newtonian physics")
markdown_table = table.get_string()
display(Markdown(markdown_table));

Adding velocities relativistically, and discrepancy with newtonian physics


| Velocity1 (c) | Velocity2 (c) | Added relativistically (c) | Discrepancy |
|-------------: |-------------: |--------------------------: |-----------: |
|        10.00% |        10.00% |                     19.80% |        1.0% |
|        20.00% |        20.00% |                     38.46% |        3.8% |
|        30.00% |        30.00% |                     55.05% |        8.3% |
|        40.00% |        40.00% |                     68.97% |       13.8% |
|        50.00% |        50.00% |                     80.00% |       20.0% |
|        70.00% |        70.00% |                     93.96% |       32.9% |
|        90.00% |        90.00% |                     99.45% |       44.8% |
|        95.00% |        95.00% |                     99.87% |       47.4% |
|        99.00% |        99.00% |                99.9949498% |       49.5% |
|        99.99% |        99.99% |                99.9999995% |       50.0% |

Notice how the discrepancy grows up to 50%, and the added velocities never exceed $c$.

## So the answer to our original question

**99.99% $c$ + 99.99% $c$ = 99.9999995% $c$**

How about that! Regardless of your velocities, you will never measure a speed greater than $c$.

## Faster speeds need more precision

Standard Python can handle about 15 decimal places of precision, which is enough for everyday calculations. But when we get close to $c$ we need more precision, potentially hundreds of decimal places, to get an accurate answer.

The rest of these notebooks use my *relativity_lib.py* library, which uses mpmath to handle arbitrary precision arithmetic. It can perform relativistic calculations to any precision required.

Here's a very quick demo:

In [4]:
# Download relativity_lib.py from your GitHub repo if not present. Important for colab.
import os

if not os.path.exists("relativity_lib.py"):
    !wget -q https://raw.githubusercontent.com/lookbusy1344/Relativity/main/Python/relativity_lib.py

import mpmath as mp
import relativity_lib as rl

# Use 100 decimal places. Standard Python can only handle about 15-17 decimal places, we are going to need more
rl.configure(100)

very_fast = mp.mpf("0.999999999") * rl.c
immensely_fast = mp.mpf("0.9999999999999999") * rl.c

very_fast_2 = rl.add_velocities(very_fast, very_fast) / rl.c
immensely_fast_2 = rl.add_velocities(immensely_fast, immensely_fast) / rl.c

print(f"Helper library: 0.999999999c * 2 = {mp.nstr(very_fast_2, 20)}")
print(f"Helper library: 0.9999999999999999c * 2 = {mp.nstr(immensely_fast_2, 40)}")

Helper library: 0.999999999c * 2 = 0.9999999999999999995
Helper library: 0.9999999999999999c * 2 = 0.999999999999999999999999999999995


## Next steps

[Time dilation and length contraction](https://github.com/lookbusy1344/Relativity/blob/main/Python/Time%20dilation.ipynb)

[Index of my Python relativity tools](https://github.com/lookbusy1344/Relativity/blob/main/Python/README.md)

Here's the wikipedia page on the topic: https://en.wikipedia.org/wiki/Velocity-addition_formula#Special_relativity